Default worker label

wderezin opened the pull request:
- implemented labels on default_worker_pool and updates

Closes #775 Feature request: have "label" property in "ibm_container_cluster" 775

This PR requires IBM-Cloud/bluemix-go#252 which updates the SDK

Reopening PR #1940 which was accidentally closed

closes #775 
closes #1221

allen-servedio commented on ibm/resource_ibm_container_cluster.go:
```diff
@@ -157,6 +157,16 @@ func resourceIBMContainerCluster() *schema.Resource {
+				Type:             schema.TypeMap,
+				Optional:         true,
+				Computed:         true,
+				ForceNew:         true,
```
Does this mean that adding labels to the default cluster will cause it to be destroyed and recreated? If so, is that required?

wderezin replied:
No required, the API support changes.

allen-servedio commented on ibm/resource_ibm_container_cluster.go:
```diff
@@ -952,6 +963,47 @@ func resourceIBMContainerClusterUpdate(d *schema.ResourceData, meta interface{})
+		if len(workerPools) > 0 && workerPoolContains(workerPools, defaultWorkerPool) {
+			poolName = defaultWorkerPool
+			poolContains = true
+		} else if len(workerPools) > 0 && workerPoolContains(workerPools, computeWorkerPool) && workerPoolContains(workerPools, gatewayWorkerpool) {
```
Why does it also require the `gatewayWorkerpool`?

wderezin replied:
This is following the pattern of the other attribute such as https://github.com/IBM-Cloud/terraform-provider-ibm/blob/78b7ac269954ade9e9ea386c0b643bf15212f2f0/ibm/resource_ibm_container_cluster.go#L940

wderezin replied:
@hkantare can you confirm that the labels are on the gatewayWorkerpool as well?

hkantare replied:
Yes when I checked form UI we can have lables on gateway worker pool also

allen-servedio commented on ibm/resource_ibm_container_cluster.go:
```diff
@@ -952,6 +963,47 @@ func resourceIBMContainerClusterUpdate(d *schema.ResourceData, meta interface{})
+					labels[k] = v.(string)
+				}
+			}
+			err = workerPoolsAPI.UpdateLabelsWorkerPool(clusterID, poolName, labels, targetEnv)
```
So if there are no labels, does this then remove the labels by sending a blank map?

wderezin replied:
Correct.

allen-servedio commented on ibm/resource_ibm_container_cluster.go:
```diff
@@ -952,6 +963,47 @@ func resourceIBMContainerClusterUpdate(d *schema.ResourceData, meta interface{})
+			err = workerPoolsAPI.UpdateLabelsWorkerPool(clusterID, poolName, labels, targetEnv)
+			if err != nil {
+				return fmt.Errorf(
+					"Error updating the labels%s", err)
```
need a space after `labels` : `labels%s"` should be `labels %s"`

allen-servedio replied:
Also, maybe I am wrong, but don't you use `%v` for errors?

wderezin replied:
This followed the pattern in the rest of the code.

hkantare commented on ibm/resource_ibm_container_cluster.go:
```diff
@@ -157,6 +157,16 @@ func resourceIBMContainerCluster() *schema.Resource {
+				Optional:         true,
+				Computed:         true,
+				ForceNew:         false,
+				DiffSuppressFunc: applyOnce,
```
As per my knowledge IKS also appneds its own labels  and when we set back in Read method it should be combination of user defined and API defien d labels

wderezin replied:
This is the same functionality as labels on an attached worker pool -- which only applies the labels once to avoid the Read issue you mentioned.

wderezin replied:
Let me see if I can add the functionality to update labels as well.

wderezin replied:
I added a IgnoreSystemLabels that runs on read for all worker pools (container default worker pool, container worker pools, vpc default worker pool, and vpc worker pools) that ignores all labels that start with `ibm-cloud.kubernetes.io/`, `kubernetes.io/` and `k8s.io`   With this I was able to remove the applyOnce requirement on the labels.

wderezin commented on ibm/resource_ibm_container_worker_pool_test.go:
```diff
@@ -30,7 +30,7 @@ func TestAccIBMContainerWorkerPool_basic(t *testing.T) {
 						"ibm_container_worker_pool.test_pool", "size_per_zone", "1"),
 					resource.TestCheckResourceAttr(
-						"ibm_container_worker_pool.test_pool", "labels.%", "3"),
+						"ibm_container_worker_pool.test_pool", "labels.%", "2"),
```
Number of labels should decrease by because the system labels are now being ignored on read.

wderezin commented:
This PR is ready for review and integration testing. 

I updated all the test cases for labels on containers and worker_pools but wasn't able to run `make testacc` due not having a test environment and setting up the 60+ required environment variables.

However, I did  test by doing the following:
- I deployed both an ibm_container_cluster and ibm_container_vpc_cluster both with attached worker_pools. 
- Set the labels on deployment
- Verified labels where set buy using the `data ibm_container_cluster` and `data ibm_container_vpc_cluster` 
- Updated all the labels
- Verified labels where set buy using the `data ibm_container_cluster` and `data ibm_container_vpc_cluster` 

@hkantare

wderezin commented:
Also, I use direnv to setup a local [go environment](https://github.com/direnv/direnv/blob/master/man/direnv-stdlib.1.md#layout-go)  for project.  I updated the `find` commands when doing `go fmt` to ignore the `.direnv` directory.
